doc/adr/0004-marconi-initiative.rst

```diff
@@ -0,0 +1,100 @@
+  
+  C. Also due to the monolithic architecture that assumes that there is only one index running we ran into trouble when we made the Plutus Application Backend collect and index information requested by smart contracts. Now we have two components that index information from the blockchain, but they are not synchronised. Querying the Chain Index about transactions received from the Plutus Application Backend often results in no data returned, since the Chain Index indexes data slower than the PAB.
+  
+  D. The lack of non-functional requirements resulted in software that uses an unreasonable amount of resources and results in slow synchronisation speeds. And since everything is monolithic it is difficult to turn off indexing of data which is not required by our customers there is no way to limit the required resources.
+  
+  E. The same lack of a specification and non-functional requirements makes the testing feel ad-hoc and like an afterthought.
+
+The Chain Index was meant to be a software application that supports the execution of smart contracts. And, in that, it succeeded. However, we found out that our customers would rather have a library of functionality that they can customise to build their indexers, only for data they care about for their application, using whatever storage engine that they prefer and supporting only the queries that they need to support.
+
+So when we took all the feedback into account we decided that a redesign of the indexing solution using a much simplified and modular design is a worthwhile enterprise.
+
+We continue by introducing some of the design principles that guided us in the specification of Marconi.
+
+============================
+Design principles of Marconi
+============================
+
+I follow the Algebra Driven Design approach for Marconi components, so from the get-go, we will have a checked specification for the software that we develop.
+
+The specification is based on a simplified model which should help with documenting how everything works without getting into the more complex details.
+
+Having a set of property-based tests to validate that the implementation conforms to the specification also means that the correctness of the implementation does not rely on type-level checks or complicated term-level machinery (we could even verify the correctness of a Rust implementation by leveraging the Rust to Haskell FFI).
+
+Because we have no reliance on type-level checks or complicated architectural patterns to validate the software (we use the specification and property tests for that), the code is much easier to understand, document and extend.
+
+=================
+Indexing solution
+=================
+
+The indexing solution has the following basic requirements: it needs to deal with rollbacks as elegantly as possible and provide a way to compromise between memory, disk and CPU usage.
```

Comment:
```suggestion
Any indexing solution has to deal with blockchain rollbacks, which can happen as the network achieves consensus. We introduce the following basic requirements: it needs to deal with rollbacks as elegantly as possible and provide a way to compromise between memory, disk and CPU usage.
```

Expanding a bit just because you haven't mentioned rollbacks before.